### ISSUE TITLE
Union of Lists vs List of Unions

### ISSUE THREAD

**Issue by samuelstevens** [python/mypy#12131]
**Bug Report**

Mypy complains about assigning a `Union[List[str], List[int]]` to a variable with type `List[Union[str, int]]`. It seems to me that `Union[List[str], List[int]]` is either a list of strings or a list of ints, and that these lists are valid values for `List[Union[str, int]]`. Perhaps there is a invariance/covariance issue I do not understand?

**To Reproduce**
```python
from typing import List, Union

def union_of_lists() -> Union[List[str], List[int]]:
    if True:
        return ["a"]
    else:
        return [1]

expanded: List[Union[str, int]] = union_of_lists()
```

```sh
mypy --strict minimal_example.py
```

**Expected Behavior**

I expected no issues found.

**Actual Behavior**

I get:

```
minimal_example.py:11: error: Incompatible types in assignment (expression has type "Union[List[str], List[int]]", variable has type "List[Union[str, int]]")
```

**Your Environment**

- Mypy version used: 0.910
- Mypy command-line flags: --strict
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.9.7
- Operating system and version: macOS

**Comment by erictraut** [comment]
The type `list[str | int]` is different from the type `list[str] | list[int]`. The former is a list that contains a mixture of `str` and `int` values. The latter is either a list that contains only `str` values or a list that contains only `int` values. These are not compatible types, so mypy is correct in flagging this as an error. The type parameter for the `list` class is invariant.

Consider the following:
```python
def mutate(x: list[str | int]):
    x.append(1)

my_str_list: list[str] = ["A", "B"]
mutate(my_str_list) # Assume that mypy allowed this

# This will crash because my_str_list now contains an int.
print([x.lower() for x in my_str_list])
```

**Comment by samuelstevens** [comment]
Ok, I suspected it was an invariance issue. I understand why the code in your example is broken, and why mypy flags it as incorrect. If you could change `mutate` to `Sequence[str | int]` (and then remove the call to append), then `list[str]` can be used as a `Sequence[str | int]`, right?

In my case, however, I am not passing my `list[str | int]` to anything function that is going to mutate it, etc. While I understand your example, I'm really struggling to see how my example code could lead to a type error.

I'm closing this issue because I don't think it's a bug, but I would really appreciate any further explanation.

**Comment by erictraut** [comment]
Yes, if you change the type declaration for `expanded` to use a `Sequence` in your example, it eliminates the type violation.

```python
expanded: Sequence[str | int] = union_of_lists()
```

This is safe because `expanded` is immutable, and the type checker will enforce the immutability.
